File: app/src/main/java/org/dhis2/data/service/ReservedValuesWorkerModule.java
package org.dhis2.data.service;

import android.app.NotificationManager;
import android.content.Context;

import androidx.annotation.NonNull;

import org.dhis2.commons.di.dagger.PerService;
import org.dhis2.commons.prefs.PreferenceProvider;
import org.dhis2.data.biometrics.BiometricsConfigApi;
import org.dhis2.data.biometrics.BiometricsConfigRepositoryImpl;;
import org.dhis2.data.service.workManager.WorkManagerController;
import org.dhis2.usescases.biometrics.BiometricsConfigRepository;
import org.dhis2.utils.analytics.AnalyticsHelper;
import org.hisp.dhis.android.core.D2;

import dagger.Module;
import dagger.Provides;

@Module
public class ReservedValuesWorkerModule {

    @Provides
    @PerService
    NotificationManager notificationManager(@NonNull Context context) {
        return (NotificationManager) context.getSystemService(Context.NOTIFICATION_SERVICE);
    }

    @Provides
    @PerService
    SyncRepository syncRepository(@NonNull D2 d2) {
        return new SyncRepositoryImpl(d2);
    }

    @Provides
    @PerService
    BiometricsConfigRepository biometricsConfigRepository(
            @NonNull D2 d2,
            @NonNull PreferenceProvider preferences
    ) {
        BiometricsConfigApi biometricsConfigApi = d2.retrofit().create(BiometricsConfigApi.class);
        return new BiometricsConfigRepositoryImpl(d2, preferences, biometricsConfigApi);
    }

    @Provides
    @PerService
    SyncPresenter syncPresenter(
            @NonNull D2 d2,
            @NonNull PreferenceProvider preferences,
            @NonNull WorkManagerController workManagerController,
            @NonNull AnalyticsHelper analyticsHelper,
            @NonNull SyncStatusController syncStatusController,
            @NonNull SyncRepository syncRepository,
            @NonNull BiometricsConfigRepository biometricsConfigRepository
    ) {
        return new SyncPresenterImpl(d2, preferences, workManagerController, analyticsHelper, syncStatusController, syncRepository,biometricsConfigRepository);
    }
}

File: app/src/main/java/org/dhis2/data/service/ReservedValuesWorkerModule.kt
package org.dhis2.data.service

import android.app.NotificationManager
import android.content.Context
import dagger.Module
import dagger.Provides
import org.dhis2.commons.di.dagger.PerService
import org.dhis2.commons.prefs.BasicPreferenceProvider
import org.dhis2.commons.prefs.PreferenceProvider
import org.dhis2.data.biometrics.BiometricsConfigApi
import org.dhis2.data.biometrics.BiometricsConfigRepositoryImpl
import org.dhis2.data.biometrics.BiometricsParentChildConfigApi
import org.dhis2.data.biometrics.BiometricsParentChildConfigRepositoryImpl
import org.dhis2.data.notifications.NotificationD2Repository
import org.dhis2.data.notifications.NotificationsApi
import org.dhis2.data.notifications.UserGroupsApi
import org.dhis2.data.service.workManager.WorkManagerController
import org.dhis2.usescases.biometrics.repositories.BiometricsConfigRepository
import org.dhis2.usescases.biometrics.repositories.BiometricsParentChildConfigRepository
import org.dhis2.usescases.notifications.domain.NotificationRepository
import org.dhis2.utils.analytics.AnalyticsHelper
import org.hisp.dhis.android.core.D2


@Module
class ReservedValuesWorkerModule {
    @Provides
    @PerService
    fun notificationManager(context: Context): NotificationManager {
        return context.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
    }

    @Provides
    @PerService
    fun syncRepository(d2: D2): SyncRepository {
        return SyncRepositoryImpl(d2)
    }

    @Provides
    @PerService
    fun biometricsConfigRepository(
        d2: D2,
        basicPreferences: BasicPreferenceProvider
    ): BiometricsConfigRepository {
        val biometricsConfigApi = d2.retrofit().create(
            BiometricsConfigApi::class.java
        )
        return BiometricsConfigRepositoryImpl(d2, basicPreferences, biometricsConfigApi)
    }

    @Provides
    @PerService
    fun biometricsParentChildConfigRepository(
        d2: D2,
        basicPreferences: BasicPreferenceProvider
    ): BiometricsParentChildConfigRepository {
        val biometricsConfigApi = d2.retrofit().create(
            BiometricsParentChildConfigApi::class.java
        )
        return BiometricsParentChildConfigRepositoryImpl(basicPreferences, biometricsConfigApi)
    }

    @Provides
    @PerService
    fun notificationsRepository(
        d2: D2,
        preference: BasicPreferenceProvider
    ): NotificationRepository {
        val notificationsApi = d2.retrofit().create(
            NotificationsApi::class.java
        )
        val userGroupsApi = d2.retrofit().create(UserGroupsApi::class.java)
        return NotificationD2Repository(d2, preference, notificationsApi, userGroupsApi)
    }

    @Provides
    @PerService
    internal fun syncPresenter(
        d2: D2,
        preferences: PreferenceProvider,
        workManagerController: WorkManagerController,
        analyticsHelper: AnalyticsHelper,
        syncStatusController: SyncStatusController,
        syncRepository: SyncRepository,
        biometricsConfigRepository: BiometricsConfigRepository,
        biometricsParentChildConfigRepository: BiometricsParentChildConfigRepository,
        notificationsRepository: NotificationRepository
    ): SyncPresenter {
        return SyncPresenterImpl(
            d2,
            preferences,
            workManagerController,
            analyticsHelper,
            syncStatusController,
            syncRepository,
            biometricsConfigRepository,
            biometricsParentChildConfigRepository,
            notificationsRepository,
        )
    }
}
